File: examples/03-structured-extraction-invoices/ts/main.ts
// Structured invoice extraction — TypeScript.
// Binds Claude to a Zod schema via the LLM's `responseFormat` option so the
// returned content is guaranteed to parse into an `Invoice` object.

import "dotenv/config";
import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { anthropic } from "@llamaindex/anthropic";
import { z } from "zod";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_FILE = resolve(__dirname, "..", "data", "invoice_001.txt");

const LineItem = z.object({
  description: z.string(),
  quantity: z.number(),
  unitPrice: z.number(),
  amount: z.number(),
});

const Invoice = z.object({
  vendorName: z.string(),
  vendorTaxId: z.string().nullable().optional(),
  invoiceNumber: z.string(),
  invoiceDate: z.string().describe("ISO-8601 date, e.g. 2026-04-18"),
  dueDate: z.string().nullable().optional(),
  poNumber: z.string().nullable().optional(),
  lineItems: z.array(LineItem),
  subtotal: z.number(),
  tax: z.number(),
  total: z.number(),
});

const SYSTEM =
  "Extract invoice fields from the user-supplied invoice text. " +
  "Preserve currency amounts and signs exactly. Include discounts and " +
  "shipping as line items. Dates must be ISO-8601.";

async function main(): Promise<void> {
  const invoiceText = readFileSync(DATA_FILE, "utf8");
  const llm = anthropic({ model: "claude-sonnet-4-6", maxTokens: 2048 });

  const response = await llm.chat({
    messages: [
      { role: "system", content: SYSTEM },
      { role: "user", content: invoiceText },
    ],
    responseFormat: Invoice,
  });

  // When responseFormat is a Zod schema, content is the JSON string conforming to it.
  const parsed = Invoice.parse(JSON.parse(String(response.message.content)));
  console.log(JSON.stringify(parsed, null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/03-structured-extraction-invoices/py/main.py
"""Structured invoice extraction — Python.

Binds Claude to a Pydantic schema using `LLM.as_structured_llm` so the
model output is guaranteed to validate against `Invoice`.
"""

from datetime import date
from pathlib import Path

from dotenv import load_dotenv
from llama_index.core.llms import ChatMessage
from llama_index.llms.anthropic import Anthropic
from pydantic import BaseModel, Field

REPO_ROOT = Path(__file__).resolve().parents[3]
DATA_FILE = Path(__file__).resolve().parents[1] / "data" / "invoice_001.txt"

load_dotenv(REPO_ROOT / ".env")


class LineItem(BaseModel):
    description: str = Field(description="Item or service description, as it appears on the invoice.")
    quantity: float = Field(description="Number of units. Use 1 for shipping/discount lines.")
    unit_price: float = Field(description="Price per unit in invoice currency.")
    amount: float = Field(description="Line total (quantity * unit_price, signed for discounts).")


class Invoice(BaseModel):
    vendor_name: str
    vendor_tax_id: str | None = None
    invoice_number: str
    invoice_date: date
    due_date: date | None = None
    po_number: str | None = None
    line_items: list[LineItem]
    subtotal: float
    tax: float
    total: float


SYSTEM = (
    "Extract invoice fields from the user-supplied invoice text. "
    "Preserve the currency amounts and signs exactly. Include discounts and "
    "shipping as line items. Dates must be ISO-8601."
)


def main() -> None:
    invoice_text = DATA_FILE.read_text()

    llm = Anthropic(model="claude-sonnet-4-6", max_tokens=2048)
    structured_llm = llm.as_structured_llm(Invoice)

    response = structured_llm.chat(
        [
            ChatMessage(role="system", content=SYSTEM),
            ChatMessage(role="user", content=invoice_text),
        ]
    )

    invoice: Invoice = response.raw  # type: ignore[assignment]
    print(invoice.model_dump_json(indent=2))


if __name__ == "__main__":
    main()
